### PR TITLE
DTD Formatting implemented

### LINKED ISSUE #221: DTD elements are not formatted
Following up on #90, given:
```
<?xml version="1.0"?>
<!DOCTYPE note [
<!ELEMENT note (to,from,heading,body)>
    <!ELEMENT to (#PCDATA)>
        <!ELEMENT from (#PCDATA)>
                <!ELEMENT heading (#PCDATA)>
            <!ELEMENT body (#PCDATA)>
]>
<note>
    <from>Jani</from>
    <heading>Reminder</heading>
    <body>Don't forget me this weekend</body>
</note>
```

Formatting the document ignores the DOCTYPE content.

### PR DESCRIPTION
**Pull request by NikolasKomonen**
Fixes #221, #250, #268, #226

### CODE REVIEW

**Comment by fbricon** [comment]
With this PR, the DTD from https://github.com/angelozerr/lsp4xml/issues/268#issue-392761320 is broken after formatting. 

it changes
 `<!ELEMENT dl (dt|dd)+>`
into 
`<!ELEMENT dl (dt|dd) +>`

The plus sign should stick to the closing parenthesis

**Comment by fbricon** [comment]
```xml
<!ATTLIST img
  src     CDATA  #REQUIRED
  alt     CDATA  #IMPLIED
  height  CDATA  #IMPLIED
  width   CDATA  #IMPLIED
  align   CDATA  #IMPLIED
  border  CDATA  #IMPLIED
  hspace  CDATA  #IMPLIED
  vspace  CDATA  #IMPLIED
  %all;>


<!-- Hypertext anchors.
	CONSTRAINT: A elements are not allowed inside
	other A elements, a fact that XML cannot express. -->
```
is turned into

```
<!ATTLIST img
   src CDATA #REQUIRED
   alt CDATA #IMPLIED
   height CDATA #IMPLIED
   width CDATA #IMPLIED
   align CDATA #IMPLIED
   border CDATA #IMPLIED
   hspace CDATA #IMPLIED
   vspace CDATA #IMPLIED
   %all;

<!-- Hypertext anchors.
	CONSTRAINT: A elements are not allowed inside
	other A elements, a fact that XML cannot express. -->
```
the closing bracket disappears. Also, there's an extra space added in front of each attribute.

screen shot 2019-01-08 at 11 16 40 am

**Comment by NikolasKomonen** [comment]
@fbricon Updated

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLFormatter.java [hunk 1]
```diff
@@ -217,6 +168,89 @@ private void format(DOMNode node, int level, int end, XMLBuilder xml) {
...
+				DOMProcessingInstruction processingInstruction = (DOMProcessingInstruction) node;
+				xml.startPrologOrPI(processingInstruction.getTarget());
+				if (node.hasAttributes()) {
+					// generate attributes
+					String[] attributes = new String[3];
+					attributes[0] = "version";
+					attributes[1] = "encoding";
+					attributes[2] = "standalone";
+
+					for (int i = 0; i < attributes.length; i++) {
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLFormatter.java]: for (String name : attributes)

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLFormatter.java [hunk 2]
```diff
@@ -217,6 +168,89 @@ private void format(DOMNode node, int level, int end, XMLBuilder xml) {
...
+						String name = attributes[i];
+						String value = node.getAttribute(attributes[i]);
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/services/XMLFormatter.java]: value = node.getAttribute(name)

File: org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/StringUtils.java [hunk 3]
```diff
@@ -83,6 +83,8 @@ public static String getStartWhitespaces(String lineText) {
 		return whitespaces.toString();
 	}
 
+
```
Comment (fbricon) [inline org.eclipse.lsp4xml/src/main/java/org/eclipse/lsp4xml/utils/StringUtils.java]: unnecessary change

**Comment by NikolasKomonen** [comment]
@fbricon tested again with dtds, all looks good.
